src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -542,37 +544,51 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
             }
 
             var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+            var nullabilityDiagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
             var underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
             TupleTypeSymbol.GetUnderlyingTypeChain(type, underlyingTupleTypeChain);
 
             int offset = 0;
             foreach (var underlyingTuple in underlyingTupleTypeChain)
             {
                 ArrayBuilder<TypeParameterDiagnosticInfo> useSiteDiagnosticsBuilder = null;
-                CheckTypeConstraints(underlyingTuple, conversions, includeNullability, currentCompilation, diagnosticsBuilder, warningsBuilderOpt: diagnosticsBuilder, ref useSiteDiagnosticsBuilder);
+                CheckTypeConstraints(underlyingTuple, conversions, includeNullability, currentCompilation, diagnosticsBuilder, nullabilityDiagnosticsBuilderOpt: nullabilityDiagnosticsBuilder, ref useSiteDiagnosticsBuilder);
 
                 if (useSiteDiagnosticsBuilder != null)
                 {
                     diagnosticsBuilder.AddRange(useSiteDiagnosticsBuilder);
                 }
 
-                foreach (var pair in diagnosticsBuilder)
+                populateDiagnostics(diagnosticsBuilder, diagnosticsOpt);
+                populateDiagnostics(nullabilityDiagnosticsBuilder, nullabilityDiagnosticsOpt);
+
+                diagnosticsBuilder.Clear();
+                nullabilityDiagnosticsBuilder.Clear();
```

Comment:
Consider moving this into `populateDiagnostics` (perhaps renamed to `populateDiagnosticsAndClear`).